packages/@aws-cdk/aws-servicecatalog/lib/private/association-manager.ts

```diff
@@ -0,0 +1,48 @@
+import * as cdk from '@aws-cdk/core';
+import { CfnResource } from '@aws-cdk/core';
+import { TagUpdateConstraintOptions } from '../constraints';
+import { IPortfolio } from '../portfolio';
+import { IProduct } from '../product';
+import { CfnPortfolioProductAssociation, CfnResourceUpdateConstraint } from '../servicecatalog.generated';
+import { hashValues } from './util';
+import { InputValidator } from './validation';
+
+export class AssociationManager {
+  public static associateProductWithPortfolio(portfolio: IPortfolio, product: IProduct): string {
+    const associationKey = hashValues(portfolio.node.addr, product.node.addr);
+    const constructId = `PortfolioProductAssociation${associationKey}`;
+
+    if (!portfolio.node.tryFindChild(constructId)) {
+      new CfnPortfolioProductAssociation(portfolio as unknown as cdk.Resource, constructId, {
+        portfolioId: portfolio.portfolioId,
+        productId: product.productId,
+      });
+    }
+    return associationKey;
+  }
+
+  public static constrainTagUpdates(portfolio: IPortfolio, product: IProduct, options: TagUpdateConstraintOptions): void {
+    InputValidator.validateLength(this.prettyPrintAssociation(portfolio, product), 'description', 0, 2000, options.description);
+    const associationKey = this.associateProductWithPortfolio(portfolio, product);
+    const constructId = `ResourceUpdateConstraint${associationKey}`;
+
+    if (!portfolio.node.tryFindChild(constructId)) {
+      const constraint = new CfnResourceUpdateConstraint(portfolio as unknown as cdk.Resource, constructId, {
+        acceptLanguage: options.messageLanguage,
+        description: options.description,
+        portfolioId: portfolio.portfolioId,
+        productId: product.productId,
+        tagUpdateOnProvisionedProduct: options.allow === false ? 'NOT_ALLOWED' : 'ALLOWED',
+      });
+
+      // Add dependsOn to force proper order in deployment.
+      constraint.addDependsOn(portfolio.node.tryFindChild(`PortfolioProductAssociation${associationKey}`) as CfnResource);
```

Comment:
We really shouldn't be duplicating this `PortfolioProductAssociation${associationKey}` construct ID. Should be extracted to be more [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself).

Maybe make `associateProductWithPortfolio()` return both the association key and the `CfnPortfolioProductAssociation`?

```ts
  public static associateProductWithPortfolio(
    portfolio: IPortfolio, product: IProduct,
  ): { associationKey: string, cfnPortfolioProductAssociation: CfnPortfolioProductAssociation } {
```

Reply:
what about just returning  a `[string, string]` of the associationKey and associationConstructId, respectively?

Reply:
oh no, returning association is better since we don't have to add the nodeChild part again in depends on.

Reply:
I think have workable solution, but want to make sure this won't be upsetting ... 
```  
public static associateProductWithPortfolio(
    portfolio: IPortfolio, product: IProduct,
  ): { associationKey: string, cfnPortfolioProductAssociation: CfnPortfolioProductAssociation } {
    const associationKey = hashValues(portfolio.node.addr, product.node.addr, portfolio.stack.node.addr);
    const constructId = `PortfolioProductAssociation${associationKey}`;
    const associationNode = portfolio.node.tryFindChild(constructId);

    let createAssociation: () => CfnPortfolioProductAssociation = function (): CfnPortfolioProductAssociation {
      return new CfnPortfolioProductAssociation(portfolio as unknown as cdk.Resource, constructId, {
        portfolioId: portfolio.portfolioId,
        productId: product.productId,
      });
    };

    return {
      associationKey: associationKey,
      cfnPortfolioProductAssociation: !associationNode ? createAssociation() : associationNode as CfnPortfolioProductAssociation,
    };
  }
```
Making it a separate function seems like a lot because we would have to pass in all the other variables again, and it's basically to get around typescript not liking (at least the way we tried) to have it create association if the nodeChild part doesn't exist.

Reply:
I don't understand... you already call `node.tryFindChild()` in that method. So you either create the `CfnPortfolioProductAssociation`, or you get if from `node.tryFindChild()`, so you always have something to return here.

Reply:
I think I was unclear, by separate function I had meant pulling it out of this one and creating another `private static ...` one.  But yes, you don't need to necessarily define anything I just wasn't sure if it would look pretty enough.  I just moved the body of the function into the return statement.